### DISCUSSION TITLE
How to delete empty sections from powerpoint file with ShapeCrawler nuget package?

### DISCUSSION THREAD

**Discussion by shubhjack** [ShapeCrawler/ShapeCrawler#239] [Ideas]
How to delete empty sections from powerpoint file with ShapeCrawler nuget package?

**Comment by AdamShakhabov** [comment]
"empty sections" is not clear. Could you provide a little bit more details? Is it an empty shape? Would be could if you attach pptx-file sample.

**Reply by shubhjack** [reply]
[Section-No-Slide.pptx](https://github.com/ShapeCrawler/ShapeCrawler/files/8637719/Section-No-Slide.pptx)
<img width="886" alt="SectionWithoutSlide" src="https://user-images.githubusercontent.com/104970659/167072388-4da5e39a-8327-4527-b0d0-0a752e10c893.PNG">
<img width="959" alt="SectionsWithSlide" src="https://user-images.githubusercontent.com/104970659/167072394-4090a847-05f5-4699-ab23-4ef297c25caa.PNG">
[Section-With-Slides.pptx](https://github.com/ShapeCrawler/ShapeCrawler/files/8637720/Section-With-Slides.pptx)

I've attached 2 screenshots and 2 PPTs for reference with self explainatory names. So basically what I want to achieve is to delete slides and if its the last slide in that section then delete that section as well. I'm able to delete slides but there is no way I'm able to access Section. I hope you will provide some suggestion or solution because I urgently need the solution. 
**Thank you so much in Advance !**

**Comment by AdamShakhabov** [comment]
Currently, section handling is not supported by ShapeCrawler. By the way, this is a good feature for future versions!

You can you use the following code sample to remove empty sections with pure [Open XML SDK](https://github.com/OfficeDev/Open-XML-SDK):

```C#
using DocumentFormat.OpenXml.Office2010.PowerPoint;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Presentation;

const string pptxPath = @"c:\temp\test.pptx";

using var doc = PresentationDocument.Open(pptxPath, true);
var sectionList = doc.PresentationPart!.Presentation.PresentationExtensionList?.Descendants<SectionList>().First();

if (sectionList == null)
{
    return; // presentation doesn't have sections
}

// Remove empty sections
foreach (Section section in sectionList)
{
    var isEmptySection = !section.Descendants<SlideId>().Any();
    if (isEmptySection)
    {
        section.Remove();
    }
}
```

I tested this code on just a few test cases. Let me know if it doesn't work for you.

**Reply by shubhjack** [reply]
I tested your code, Its working but actually deleting the first section because that condition is not working as expected. Its giving false even though section has slides and deleting that section. I changed that condition but now facing one problem which is as soon as empty section is deleted the debugger is coming out of foreach loop instead of checking for next items (sections) in **sectionList**.


`//var isEmptySection = !section.Descendants<SlideId>().Any();`
`var isEmptySection = section.SectionSlideIdList.Count() == 0 ? true : false;`

**Comment by shubhjack** [comment]
So I was able to achieve the requirement, here is the complete API. **Thanks for your help !**
**Note:** deletion is not working if ppt is opened in readonly mode.

```
        [HttpGet]
        [Route("[action]")]
        public string DeleteSections()
        {
            int count = 0;
            
            // Open the presentation as edit-mode.
            using (PresentationDocument presentationDocument = PresentationDocument.Open("PPT.pptx", true))
            {
                var sectionList = presentationDocument.PresentationPart!.Presentation.PresentationExtensionList?.Descendants<SectionList>().First();

                if (sectionList == null)
                {
                    return "No sections"; // presentation doesn't have sections
                }

                List<Section> emptySectionList = new List<Section>();

                // Create list of empty sections
                foreach (Section section in sectionList)
                {
                    var isEmptySection = section.SectionSlideIdList.Count() == 0 ? true : false;

                    if (isEmptySection)
                    {
                        emptySectionList.Add(section);
                        count++;
                    }
                }

                // Remove empty sections
                foreach (var item in emptySectionList)
                {
                    sectionList.RemoveChild(item);
                }
            }
            return $"{count} empty sections deleted";
        }
```

**Comment by shubhjack** [comment]
@adamshakhabov - Is there any way I can have a MemoryStream of updated PPT from OpenXML ? I want to upload it in sharepoint but when I'm opening the file from MemoryStream and after deleting empty sections when I'm saving it my changes are not there.
Could you please help me out ?
Method is same as above just doing the same with stream
```
        [HttpGet]
        [Route("[action]")]
        public IActionResult RemoveEmptySection(IFormFile file)
        {
            int count = 0;
            //using (MemoryStream stream = new MemoryStream(ConvertToStream(file)))
            //{
                var stream = new MemoryStream(ConvertToStream(file));
                using (PresentationDocument presentationDocument = PresentationDocument.Open(stream, true))
                {
                    var sectionList = presentationDocument.PresentationPart!.Presentation.PresentationExtensionList?.Descendants<SectionList>().First();

                    List<Section> emptySectionList = new List<Section>();

                    // Remove empty sections
                    foreach (Section section in sectionList)
                    {
                        var isEmptySection = section.SectionSlideIdList.Count() == 0 ? true : false;

                        if (isEmptySection)
                        {
                            emptySectionList.Add(section);
                            count++;
                        }
                    }

                    foreach (var item in emptySectionList)
                    {
                        sectionList.RemoveChild(item);
                    }
                    presentationDocument.PresentationPart.Presentation.Save();

                    var mimeType = "application/vnd.openxmlformats-officedocument.presentationml.presentation";

                    return new FileContentResult(stream.ToArray(), mimeType)
                    {
                        FileDownloadName = "NewFile.pptx"
                    };
                }

                
            //}
            //return Ok($"{count} empty sections deleted");
        }
```

**Reply by AdamShakhabov** [reply]
Call `presentationDocument.Close()` instead of `presentationDocument.PresentationPart.Presentation.Save()`. This correctly applies your changes to source stream.

By the way, it is strange, even `presentationDocument.Save()` doesn't save changes. Looks like it is a bug in Open XML SDK🤔

**Reply by shubhjack** [reply]
Thanks man ! It worked ! You are a savior !
Your nuget package is simple to use and this OpenXml is just the opposite. 
Is there a way to overcome maximum size limit and slide limit in ShapeCrawler ? currently it is 250 slides and 150 Mb file size. 
**My requirement** - Delete slides using slide indexes, keeping sections intact and if all slides from a section is deleted, delete that section as well. Max file size 250 Mb is required (don't ask why :D)

**Reply by AdamShakhabov** [reply]
Yes, we can increase these limits. Max file size 250 Mb. What about slides number, 300 slides per presentation, is it ok?

To be honest, there is not some concrete reason why these limitations are presented in the library.

**Reply by shubhjack** [reply]
If you can increase the limit its a bonus. Thanks

**Comment by AdamShakhabov** [comment]
Presentation sections handling was added in [**v0.30.0**](https://github.com/ShapeCrawler/ShapeCrawler/releases/tag/v0.30.0) release.

**Reply by shubhjack** [reply]
Thanks for the update !

**Reply by shubhjack** [reply]
@adamshakhabov - Count of slides under section is showing as 0 for all sections

**Reply by AdamShakhabov** [reply]
Thank you for the report. I was able to reproduce this. _Fixing..._

**Reply by AdamShakhabov** [reply]
@shubhjack, please check this build [**ShapeCrawler.0.30.1-beta.1**](https://github.com/ShapeCrawler/ShapeCrawler/files/8770294/ShapeCrawler.0.30.1-beta.1.nupkg.zip)

**Reply by shubhjack** [reply]
Hey @adamshakhabov, I tested the beta version, Remove section is working fine independently but when I'm trying to remove slides from Slide list the Sections object throws exception so I'm not able to delete empty sections after slide removal. I was trying to use that IsRemoved property as well but modifying it is also not helping. 
Check the below service method for my api-

```
        public async Task<byte[]> RemoveSlidesAndEmptySections(string Id, int[] SlidesSelected)
        {
            byte[] fileBytes = await gridFSBucket.DownloadAsBytesAsync(new ObjectId(Id), null);
            MemoryStream stream = new MemoryStream(fileBytes);

            using (IPresentation FinalPPT = SCPresentation.Open(stream, isEditable: true))
            {
                int j = -1;
                foreach (ISlide slide in FinalPPT.Slides)
                {
                    j++;
                    if (!SlidesSelected.Contains(j))
                    {
                        FinalPPT.Slides.Remove(slide);
                    }
                }
                FinalPPT.Save();
                foreach (ISection section in FinalPPT.Sections)
                {
                    if (section.Slides.Count == 0)
                    {
                        FinalPPT.Sections.Remove(section);
                    }
                }
                FinalPPT.Save();
            }
            return stream.ToArray();
        }
```

**Reply by shubhjack** [reply]
![image](https://user-images.githubusercontent.com/104970659/170683659-cb3aefd0-839f-4051-8043-864fff139ae3.png)

Stacktrace
at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
   at ShapeCrawler.Collections.SCSlideCollection.GetBySlideId(String slideId)
   at ShapeCrawler.SCSectionCollection.Create(SCPresentation presentation)
   at ShapeCrawler.SCPresentation.get_Sections()

**Reply by AdamShakhabov** [reply]
I was able to reproduce the bug. Fixing...

**Reply by AdamShakhabov** [reply]
Fixed. @shubhjack, please check this build [**ShapeCrawler.0.30.1-beta.2**](https://github.com/ShapeCrawler/ShapeCrawler/files/8790919/ShapeCrawler.0.30.1-beta.2.nupkg.zip)

**Reply by shubhjack** [reply]
@adamshakhabov - working perfectly now !

**Reply by shubhjack** [reply]
@adamshakhabov - I was trying to optimize the above code but I'm unable to remove slides from section. Also I was not sure if removing slide from sections will also remove the slide from slidelist. Ideally as a end user I would like if I'm removing any slide from slide list it should be removed from sections slide list as well and vice versa. 

```
public async Task<byte[]> RemoveSlidesAndEmptySections(string Id, int[] SlidesIndexes)
        {
            byte[] fileBytes = await gridFSBucket.DownloadAsBytesAsync(new ObjectId(Id), null);
            MemoryStream stream = new MemoryStream(fileBytes);
            using (IPresentation FinalPPT = SCPresentation.Open(stream, isEditable: true))
            {
                int j = -1;
                foreach (ISection section in FinalPPT.Sections)
                {
                    j++;
                    foreach (ISlide slide in section.Slides)
                    {
                        if (!SlidesIndexes.Contains(slide.Number - 1))
                        {
                            FinalPPT.Sections[j].Slides.Remove(slide);
                        }
                    }
                    if (FinalPPT.Sections[j].Slides.Count == 0)
                    {
                        FinalPPT.Sections.Remove(section);
                    }
                }
                FinalPPT.Save();
            }
            return stream.ToArray();
        }
```

**Reply by AdamShakhabov** [reply]
@shubhjack, please check this version [**v0.31.0**](https://github.com/ShapeCrawler/ShapeCrawler/releases/tag/v0.31.0)

I a little bit changed the design of removing slides from a section. Now you should do that via presentation slide collection. Check sample in [**test**](https://github.com/ShapeCrawler/ShapeCrawler/blob/a13027c645db2c7c4331457a124a7d84d86b1362/ShapeCrawler.Tests/PresentationTests.cs#L190).
